\section{The Optimized Workflow}
The complexity of MRI processing workflows should be manageable to prospective users with only cursory programming experience.
It is however also important that workflow transparency, sustainability, and reproducibility are not compromised for trivial features.
We thus follow  a set of design guidelines:
(1) each workflow is represented by a high-level function, whose parameters correspond to operator-understandable concepts (i.e. describe the operations performed, rather than the computational implementation);
(2) workflow functions are highly parameterized but include workable defaults (so that users \textit{can} change their function to a significant extent without needing to edit the constituent code);
(3) graphical or interactive interfaces are avoided (as they impede reproducibility, encumber the dependency graph, and reduce the sustainability of the project).

\begin{figure*}[h!]
	\begin{subfigure}{.64\textwidth}
		\centering
		\includedot[width=\textwidth]{data/generic}
		\vspace{1.4em}
		\caption{
			“SAMRI Generic”  workflow, based on the \textcolor{mg}{\texttt{antsRegistration}} function.
			The pipeline uses a higher-resolution structural scan intermediary for registration (note the two processing streams), which facilitates differential handling of anatomical variation and susceptibility artefacts.
			The function used is highly parameterized, and both of its instances --- “s\niceus register” and “f\niceus register” --- are supplied in the workflow with defaults optimized for mouse brain registration.
			}
		\label{fig:wfgg}
	\end{subfigure}\hfill
	\begin{subfigure}{.34\textwidth}
		\centering
		\includedot[width=\textwidth]{data/legacy}
		\vspace{-1.9em}
		\caption{
			“SAMRI Legacy” workflow, which is based on the \textcolor{mg}{\texttt{antsIntroduction.sh}} function (and other functions with hard-coded parameters optimized for human brain registration), and also performs destructive affine manipulations.
			}
		\label{fig:wfgl}
	\end{subfigure}
	\caption{
		Directed acyclic graphs depicting the two alternate MRI registration workflows and their constituent processing steps.
		The package correspondence of each processing node is appended in parantheses to the node name.
		The “utility” indication corresponds to nodes based on Python functions specific to the workflow, distributed alongside it, and dynamically wrapped via Nipype.
		The “extra\niceus interfaces” indication corresponds to nodes using explicitly defined Nipype-style interfaces, which are specific to the workflow and distributed alongside it.
		}
	\label{fig:wfg}
\end{figure*}

The language of choice for workflow handling is Python, owing to its Free and Open Source (FOSS) dependency stack, readability, wealth of available libraries, ease of package management, and its large and dynamic developer community.
While workflow functions are written in Python, we also provide automatically generated Command Line Interfaces (CLIs) for use directly with Bash.
These autogenerated CLIs ensure that features become available in Bash and Python synchronously, and workflows behave identically regardless of the language in which they are invoked.

\subsection{Technologies}

Internally, the workflow functions make use of the Nipype \cite{nipype} package, which provides high-level workflow management and execution features.
Via this package, functions provided by any other package can be encapsulated in a node (complete with error reporting and isolated re-execution support) and integrated into a directed workflow graph.
Paralellization can also be managed via a number of execution plugins, allowing excellent scalability.
Most importantly, Nipype can generate graph descriptor language (DOT) summaries as well as visual representations of the workflows suitable for operator inspection, graph theoretical analysis, and programmatic comparison between workflow variants.

Via Nipype we utilize basic MRI preprocessing functions from the FSL package \cite{fsl} and registration functions from the ANTs package \cite{ants}.
While there is theoretically no limit to the number of external packages usable with Nipype, we constrain our choice as much as possible in order to minimize the dependency graph.
The choice of the ANTs package (in addition to FSL, which also provides registration functions) owes to the package providing registration functions which are more highly parameterized.
This feature allows us to avoid maladaptive optimization choices, and instead adapt the registration to the brain images at hand.
Additionally, we have implemented a number of functions in our workflow directly, e.g. to read BIDS \cite{bids} inputs, and perform dummy scans management.

Given the aforementioned guiding principles, and the hitherto listed technologies, we have constructed two registration workflows: The “Legacy” workflow (\cref{fig:wfgl}), which exhibits the common practices detailed in the \nameref{sec:bg}~section; and our novel “Generic” workflow (\cref{fig:wfgg}).
Both workflows start by performing dummy scan correction on the functional MRI data and the stimulation events file, based on automatically parsed Bruker ParaVision metadata.
The “Generic” workflow subsequently applies slice timing correction, computes the temporal mean of the functional scan (to obtain a more representative contrast for the whole time course), and applies a bias field correction to the temporal mean --- using the \textcolor{mg}{\texttt{N4BiasFieldCorrection}} function of the ANTS package, with spatial parameters adapted to the mouse brain.
Analogous operations are performed on the structural scan, following which the structural scan is registered to the reference template, and the functional scan temporal mean is registered to the structural scan --- using the \textcolor{mg}{\texttt{antsRegistration}} function of the ANTS package, with spatial parameters adapted to the mouse brain.
The structural-to-template and functional-to-structural transformation matrices are then merged, and applied in one warp computation step to the functional timecourse data --- while the structural data is warped solely based on the structural-to-template transformation matrix. 
The “Legacy” workflow follows up on dummy scan correction with a tenfold multiplication of the voxel size (making the brain size more human-like), and subsequently deletes the orientation information from the affine.
Further, the dimensions are swapped so that the data matrix matches the RPS (left$\rightarrow$Right, anterior$\rightarrow$Posterior, inferior$\rightarrow$Superior) orientation of the “Legacy” template (see \cref{fig:amb}).
Following these data manipulation steps, a temporal mean is computed, and an empirically determined signal threshold (\SI{10}{\percent} of the 98\textsuperscript{th} percentile) is applied.
Subsequently, the bias field is corrected using the \textcolor{mg}{\texttt{fast}} function of the FSL package, and parts of the image are masked using the \textcolor{mg}{\texttt{bet}} function from FSL. 
The image is then warped into the template space using the \textcolor{mg}{\texttt{antsIntroduction.sh}} function of the ANTS package, and lastly the affine variants are harmonized.

For Quality Control we distribute as part of this publication additional workflows using the NumPy \cite{numpy}, SciPy \cite{scipy}, and pandas \cite{pandas} packages, as well as Statsmodels \cite{statsmodels} for top-level statistics and matplotlib \cite{matplotlib} and Seaborn \cite{seaborn} for plotting.
Specifically, distribution densities for plots are drawn using the Scott bandwidth density estimator \cite{Scott1979}.

% Markus Marks !!! add short technology paragraph for optimization methods.

\subsection{Distribution}

As registration is a crucial step of a larger data analysis process (rather than an analysis process in its own right), the workflows are best distributed as part of a full stack (i.e. from raw data to statistical summaries) workflow package.
We include the aforementioned Generic and Legacy workflows in the SAMRI (Small Animal Magnetic Resonance Imaging) data analysis package \cite{samri} of the ETH/UZH Institute for Biomedical Engineering.

\subsection{Template Package}

\py{pytex_subfigs(
	[
		{'script':'scripts/dsurqec.py', 'label':'dsu', 'conf':'article/template.conf', 'options_pre':'{.48\\textwidth}',
			'caption':'The “Generic” template, which exemplifies $T_2$ contrast, a canonical MR \\textit{and} stereotactic data matrix orientation, a standard header with an RAS orientation, and a realistic affine transformation.
				Note the origin at Bregma which provides histologically meaningful coordinates.'
			,},
		{'script':'scripts/ambmc.py', 'label':'amb','conf':'article/template.conf', 'options_pre':'{.48\\textwidth}',
			'caption':'The “Legacy” template, which exemplifies histological contrast, the canonical histological template data matrix orientation (shared e.g. by the Allen Brain Institute template), alongside a non-standard header with features such as an RPS orientation and inflated affine transformation.'
			,},
		],
	caption='Multiplanar depcitions of the mouse brain templates, with slice coordinates centered at zero on all axes.',
	label='fig:t',)}

The suitability of a registration workflow as a standard is contingent on the quality of the template being used.
Particularly size and orientation of the template may pose constraints on the workflow.
For example, an unrealistically inflated template size mandates according parameters for all functions which deal with the data in its affine space.
Additionally, if the template axis orientation deviates by more than \SI{45}{\degree} from the image to be registered, or if an axis is flipped, the global maximum of the first (rigid) registration steps may not be correctly determined, and the image would then be skewed and nonlinearly deformed to match  the template at an incorrect orientation.
Consequently, template quality needs to be ascertained, and a workflow-compliant default should be provided.

Our recommended template (\cref{fig:dsu}) is derived from the DSURQE template --- of the Toronto Hospital for Sick Children Mouse Imaging Center \cite{dsu}.
The geometric origin of this derived template corresponds to the Bregma landmark, and provides integration with histological atlases and surgical procedures, which commonly use Bregma as a reference.
The template is in the canonical orientation of the NIfTI format, RAS (left$\rightarrow$Right, posterior$\rightarrow$Anterior, inferior$\rightarrow$Superior), and has a coronal slice positioning reflective of both the typical animal head position in MR scanners and in stereotactic surgery frames.
The template is provided at \SI{40}{\micro\meter} and \SI{200}{\micro\meter} isotropic resolutions,
and all of its associated mask and label files are identified with the prefix \textcolor{mg}{\texttt{dsurqec}} in the template packages.

We bundle the aforementioned MR template with two additional histological templates, derived from the Australian Mouse Brain Mapping Consortium (AMBMC) \cite{amb}, and the Allen Brain Institute (ABI) \cite{abi} templates.
While these atlases suffer from shortcomings listed under the \nameref{sec:bg}~section, we include the AMBMC template due to its extra long rostrocaudal coverage, and the ABI atlas due to its role as the reference atlas for numerous gene expression and projection maps.
We reorient the AMBMC template from its original RPS orientation to the canonical RAS, and apply an RAS orientation to the orientation-less ABI template after converting it to NIfTI from its original NRRD format.
These atlases are also made available at \SI{40}{\micro\meter} and \SI{200}{\micro\meter} isotropic resolutions, and the corresponding files are prefixed with \textcolor{mg}{\texttt{ambmc}} and \textcolor{mg}{\texttt{abi}}, respectively.

Additionally, we provide templates in the historically prevalent but incorrect, RPS orientation, and with the historically prevalent tenfold increase in voxel size.
These templates are derived from the DSURQE and AMBMC templates, and are prefixed with \textcolor{mg}{\texttt{ldsurque}} and \textcolor{mg}{\texttt{lambmc}}, respectively.

Lastly, due to data size considerations, we distribute \SI{15}{\micro\meter} isotropic versions of all atlases available at this resolution (AMBMC and its legacy derivative, as well as ABI) in a separate package.
The two packages which we thus distribute are called \textcolor{mg}{\texttt{mouse-brain-atlases}} and \textcolor{mg}{\texttt{mouse-brain-atlasesHD}};
and up-to-date versions of these archives can be reproduced via a FOSS script collection (which handles download, reorienting, and resampling, and was written and released for the purpose of this publication \cite{atlases_generator}).

For the comparisons performed in this text, the \textcolor{mg}{\texttt{dsurqec}} and \textcolor{mg}{\texttt{ldsurqec}} template variations (containing the same data matrix, but matched to the orientation and size requirements of the functions in the \cref{fig:wfgg} and \cref{fig:wfgl} workflows, respectively) are referred to as the “Generic” template.
Analogously, the \textcolor{mg}{\texttt{ambmc}} and \textcolor{mg}{\texttt{lambmc}} template variations are referred to as the “Legacy” template.

\subsection{Testing Dataset}

\py{pytex_tab('scripts/stim_table.py',
		label='stim',
		caption='Stimulation protocol, as delivered during functional scans.
			Stimulus event spacing and parameters are constant across scans, but exact onset time is variable in the \SI{10}{\second} magnitude range due to scanner adjustment time variability.',
		options_pre='\\scriptsize \\centering \\resizebox{\\columnwidth}{!}{',
		data='data/JogB.tsv',
		options_post='}',
		)}


For the quality control of the workflows, a dataset with an effective size of 102 scans is used.
Data from 11 adult animals is included, each animal being recorded on up to 5 sessions (repeated at 14 day intervals).
Each session contains an anatomical scan and two functional scans --- with Blood-Oxygen Level Dependent (BOLD) \cite{Ogawa1990} and Cerebral Blood Volume (CBV) \cite{Marota1999} contrast, respectively (for a total of 68 functional scans).

Anatomical scans are acquired via a TurboRARE sequence, with a RARE factor of 8, an echo time (TE) of \SI{7}{\milli\second} and a repetition time (TR) of \SI{2500}{\milli\second}, sampled at a sagittal resolution of $\mathrm{\Delta x(\nu)=\SI{166.7}{\micro\meter}}$, a horizontal resolution of $\mathrm{\Delta y(\phi)=\SI{75}{\micro\meter}}$, and a coronal resolution of $\mathrm{\Delta z(t)=\SI{650}{\micro\meter}}$ (slice thickness of \SI{500}{\micro\meter}).
The functional BOLD and CBV scans are acquired with a flip angle of \SI{60}{\degree} and with $\mathrm{TR/TE = \SI{1000}{\milli\second}/\SI{15}{\milli\second}}$ and $\mathrm{TR/TE = \SI{1000}{\milli\second}/\SI{5.5}{\milli\second}}$, respectively.
Functional scans are sampled at $\mathrm{\Delta x(\nu)=\SI{312.5}{\micro\meter}}$, $\mathrm{\Delta y(\phi)=\SI{281.25}{\micro\meter}}$, and $\mathrm{\Delta z(t)=\SI{650}{\micro\meter}}$ (slice thickness of \SI{500}{\micro\meter}).
All aforementioned scans are acquired with a Bruker PharmaScan system (\SI{7}{\tesla}, \SI{16}{\centi\meter} bore), and an in-house T/R coil.

The measured animals are fitted with an optic fiber implant ($\mathrm{l=\SI{3.2}{\milli\meter} \ d=\SI{400}{\micro\meter}}$) targeting the Dorsal Raphe (DR) nucleus in the brain stem.
The nucleus is rendered sensitive to optical stimulation by transgenic expression of Cre recombinase under the ePet promoter \cite{Scott2005} and viral injection of rAAVs delivering a plasmid with Cre-conditional expression of Channelrhodopsin and YFP ---
pAAV-EF1a-double floxed-hChR2(H134R)-EYFP-WPRE-HGHpA, a gift from Karl Deisseroth (Addgene plasmid \#20298).

The DR is stimulated via an Omicron LuxX 488-60 laser (\SI{488}{\nano\meter}) tuned to \SI{30}{\milli\watt} at contact with the fiber implant, according to the protocol listed in \cref{tab:stim}.
The operation and stimulation procedure, as well as general picture of obtained activation is consistent with previous research \cite{Grandjean2019}, and is not further commented in this study.

\subsection{Reproducibility}

The source code for this document and all data analysis shown herein (including registration and QC workflow execution) is published according to the RepSeP specifications \cite{repsep}.
The data analysis execution and document compilation has been tested repeatedly on numerous machines, and as such we attest that all figures and statistics presented can be reproduced based solely on the raw data, dependency list, and analysis scripts which we distribute.
